Check for token minting or burning

ffakenz opened the pull request:
## Why

We want to close the identified security gap where we want to prevent minting or burning of tokens in state transitions that are not actually supposed to burn or mint.

## How
🍦 Update `checkClose`, `checkContest`, `checkCommit` and `checkCollectCom` to verify no head token is burnt in v-head.

To check before merging:
* [x] CHANGELOG is up to date
* [x] ~Documentation is up to date~

v0d1ch committed:
Check for token burning in checkClose

ffakenz committed:
Add mutation to burn head token during contest

ffakenz committed:
Check for token burning in checkContest

ffakenz committed:
Add mutation to burn head token during collectcom

ffakenz committed:
Check for token burning in collectCom

ffakenz committed:
Update changelog

v0d1ch committed:
Correct haddock

pgrange commented on hydra-node/test/Hydra/Chain/Direct/Contract/Close.hs:
```diff
@@ -150,6 +150,8 @@ data CloseMutation
   | MutateCloseContestationDeadline
   | MutateCloseContestationDeadlineWithZero
   | MutateHeadId
+  | -- | Burning of the tokens should not be possible in v_head a part from 'checkAbort' or 'checkFanout'
```
```suggestion
  | -- | Burning of the tokens should not be possible if v_head is spent in an abort or fanout transaction
```

v0d1ch replied:
> Burning of the tokens should not be possible if v_head is spent in an abort or fanout transaction

Did you mean reverse here? _Burning of the tokens should not be possible **unless** v_head is spent in an abort or fanout transaction_ ?

pgrange replied:
Oups! My bad. Then my suggestion would be:
```suggestion
  | -- | Burning of the tokens should not be possible in v_head apart from 'checkAbort' or 'checkFanout'
```

pgrange commented on hydra-node/test/Hydra/Chain/Direct/Contract/Contest.hs:
```diff
@@ -220,6 +223,8 @@ genContestMutation
                   (Just $ toScriptData healthyClosedState)
               ]
+      , SomeMutation (Just "burning is forbidden") MutateTokenBurning
+          <$> changeMintedTokens tx (valueFromList [(AssetId testPolicyId "badTokenBurned", Quantity (-1))])
```
Mutate the transaction to include the minting policy as the phase-1 validation (not exercised by these tests) would not allow a transaction trying to burn/mint and which would not include the minting policy script

v0d1ch replied:
Sorry, I don't understand what exactly do you mean here? Let's pair on it?

pgrange replied:
My understanding is that you wouldn't be able to burn or mint in a transaction without providing the corresponding minting policy script. And this test, to the best of my knowledge, does not include this policy in the transaction so I feel like the transaction you test would not work anyway because of this missing.

We can pair on that tomorrow morning 👍 just ping me when you want

ch1bo replied:
You are right @pgrange! If it is hard to do that, however, I think it's fine if only the phase-2 check is tested here.

ch1bo replied:
Furthermore, we would need to have this token to burn it. So the value is in balance. Obviously the transaction inputs do not contain a "badTokenBurned" NFT

pgrange replied:
If it's hard maybe I would add some comment explaining why we do not include the minting policy eventhough we should in real life.

pgrange commented on hydra-node/test/Hydra/Chain/Direct/Contract/CollectCom.hs:
```diff
@@ -212,6 +215,8 @@ genCollectComMutation (tx, _utxo) =
         (txIn, HealthyCommit{cardanoKey}) <- elements $ Map.toList healthyCommits
         pure $ ChangeInput txIn (toUTxOContext $ mkInitialOutput testNetworkId testPolicyId cardanoKey) Nothing
+    , SomeMutation (Just "burning is forbidden") MutateTokenBurning
+        <$> changeMintedTokens tx (valueFromList [(AssetId testPolicyId "badTokenBurned", Quantity (-1))])
```
Mutate the transaction to include the minting policy as the phase-1 validation (not exercised by these tests) would not allow a transaction trying to burn/mint and which would not include the minting policy script

pgrange reviewed (approved):
This review tries to apply the [perfection game protocol](https://liveingreatness.com/core-protocols/perfection-game/)

I will describe what I like about this P.R. and then what improvement I can think of to make it perfect. I'm expecting the author of the P.R. to ensure that they understand what I write or ask for clarifications if it's not the case. I'm expecting the author of the P.R. to just not include my ideas if they disagree with them without any form of justification. Hence I, in advance, approve the P.R.

To describe what I would change, I will try to extensively use the _suggest change_ feature of GitHub as I believe this is the most efficient way to upload data from my brain to the authors' brains. That being said, the suggestions I will write will probably not compile or pass the test so do not just accept them but consider them for what they are: the description of an improvement to make the P.R. perfect.

What I like about this P.R.:
* it's been done in a very short time with a very small and precise scope in mind
* I like the name chosen for the new added functions which are explicit and clear to me

For me to find it perfect you would have to:
* make the tests pass
* polish the commit history to have only one update changelog commit
* polish the commit history to have only one haddock commit
* polish the commit history to include the changes of the one named minor into another commit
* detail the P.R. description adding at least a why section which would explain why we need to check that in the validator, in particular, explain how one could forge a transaction which would burn token in a close transaction
* see other improvement proposals below

ch1bo commented on CHANGELOG.md:
```diff
@@ -61,6 +61,8 @@ changes.
 
 -  `hydra-cluster` executable can be used to provide a local cardano "network" with `--devnet` argument
 
+- Check no tokens are burnt in v-head for close, contest and collectCom tx.
```
We have a changelog section about scripts.

Could: move this there

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Close.hs:
```diff
@@ -209,6 +211,8 @@ genCloseMutation (tx, _utxo) =
                 (Just $ toScriptData healthyCloseDatum)
             ]
+    , SomeMutation (Just "burning is forbidden") MutateTokenBurning
+        <$> changeMintedTokens tx (valueFromList [(AssetId Fixture.testPolicyId "badTokenBurned", Quantity (-1))])
```
Should: not only test burning, but also minting should not be allowed. In fact, any minting (not only of our policy). So we could generate maybe three cases:
* Burn a single token of our policy (what you did here)
* Mint a single token of our policy
* Burn or mint an `arbitrary` token of an `arbitrary` policy

This (generator) should cover the interesting edge cases with high enough frequency to get us convinced this actually works (only allows keeping `mint = 0`)

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Mutation.hs:
```diff
@@ -591,6 +592,20 @@ changeMintedValueQuantityFrom tx exclude =
 
+-- | A 'Mutation' that changes the forged quantity of tokens like this:
+-- - when no value is being forged -> add some value to be forged
+-- - when tx is forging values -> add more values on top of that
```
"Forged" is not the term we (or the cardano api/ledger etc.) would use.. it's called the mint field which can have positive (minting) or negative (burning) quantities.

ch1bo replied:
I too would have like "forge" better though.

v0d1ch replied:
I though _forge_  is more general term that covers both minting and burning but I am probably wrong. Changed to mint/burn.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Mutation.hs:
```diff
@@ -591,6 +592,20 @@ changeMintedValueQuantityFrom tx exclude =
+      TxMintValue v _ ->
+        pure $ v <> mintValue
+ where
+  mintedValue = txMintValue $ txBodyContent $ txBody tx
```
This is a very generic interface. Maybe `mintToken` or `burnToken` is what you would like to express more often in our tests? The arguments of such functions would be `PolicyId -> TokenName -> Natural` and it would be clear that we are burning 1 token when giving `1`, where in fact the quantity `-1` would be added to the `Value` in the field.

v0d1ch replied:
I added a generator for the values that should produce either mined or burned one in the `genMintedOrBurnedValue`. If we still want to be more specific I could use `Quantity` as the argument here to be explicit.

ch1bo commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -168,7 +168,8 @@ checkCollectCom ::
   Bool
 checkCollectCom ctx@ScriptContext{scriptContextTxInfo = txInfo} (contestationPeriod, parties, headId) =
-  mustContinueHeadWith ctx headAddress expectedChangeValue expectedOutputDatum
+  mustNotBurnTokens ctx headId
```
Must: also check this is in the `commit` transaction

ch1bo commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -516,6 +519,16 @@ hasPT headCurrencySymbol txOut =
+      traceIfFalse "burning is forbidden" $ Map.empty == tokenMap
+ where
+  minted = getValue (txInfoMint scriptContextTxInfo)
+{-# INLINEABLE mustNotBurnTokens #-}
```
The specification currently forbids **any** minting or burning. This implementation here only checks the minting of our currency id.

Must: change to disallow any minting or burning.

ch1bo reviewed (changes requested):
Commit transaction also must not mint/burn tokens. Also we want to prevent **any** minting or burning (per the specification).

ch1bo committed:
Do not allow any minting or burning

Values are a bit weird to work with because "no minting" actually yields
a Value with the ada symbol and quantity 0 always. This is due to the
way the ledger converts babbage tx bodies to plutusv2 script contexts.

v0d1ch committed:
Add minting/burning check when checking commits

- Rename Mutation constructors MutateTokenBurning -> MutateTokenMintingOrBurning
- Move mustNotMintOrBurn to the Util module
- Fix haddock strings to reflect also minting

v0d1ch committed:
Generate better values

- Move changelog entry and fix wording
- Fix haddock to not include the word _forge_

ffakenz committed:
minor rebase fix

ch1bo commented on CHANGELOG.md:
```diff
@@ -25,8 +25,12 @@ changes.
     were already present in Closed datum.
+=======
+  + Check no tokens are minted/burnt in v-head for close, contest, commit and collectCom tx.
+>>>>>>> 67e6f2daf (Generate better values)
```
MUST: not commit rebase conflicts

ch1bo commented on CHANGELOG.md:
```diff
@@ -63,6 +67,7 @@ changes.
 
 -  `hydra-cluster` executable can be used to provide a local cardano "network" with `--devnet` argument
 
+
```
Could: remove redundant newline

ch1bo commented on hydra-node/hydra-node.cabal:
```diff
@@ -305,6 +305,7 @@ test-suite tests
     , base
     , base16-bytestring
     , bytestring
+    , cardano-api
```
Must: not depend on cardano-api directly

Everything is re-exported and extended through hydra-cardano-api

v0d1ch replied:
This is not needed in the end.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Mutation.hs:
```diff
@@ -728,3 +746,19 @@ replaceHeadId headId = \case
   otherState -> otherState
+
+-- | Generates value such that:
+-- - alters between policy id we use in test fixtures with a random one.
```
I think that would be a reason to not keep this function "in the framework module". It's specific to our tests / fixtures.

v0d1ch replied:
I added a dedicated module for contract generators.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Mutation.hs:
```diff
@@ -728,3 +746,19 @@ replaceHeadId headId = \case
+      ]
+  tokenName <- oneof [arbitrary, pure (AssetName $ fromBuiltin hydraHeadV1)]
+  -- simulate minting or burning. 3 tokens should suffice
+  quantity <- oneof (pure <$> [-3, -2, -1, 1, 2, 3])
```
Three is quite arbitrary and cannot be controlled from the outside.

Could: use a sized generator like `arbitrarySizedIntegral` (which is the same as `arbitrary :: Int` for example) and excluding `0` using `suchThat` (it's okay if there are clearly more cases that are generated)

That way, quickcheck does test with -1 and 1 _very likely_ and random values up to size. Size is parameter which starts small and grows bigger when more quickcheck tests are run.

v0d1ch replied:
I wanted to avoid calling `suchThat` but I guess in this case it is fine since we are excluding just one value so it should not choke too much.

ch1bo reviewed (changes requested):
The mint/burn generator is good enough. Location could be improved. But address the MUST comments. Also, tests don't pass anymore as the script is now 100Bytes bigger again.

v0d1ch committed:
Fix changelog rebase mess and remove cardano-api from deps

v0d1ch committed:
Move genForParty to dedicated generator module

ch1bo commented on hydra-node/hydra-node.cabal:
```diff
@@ -262,6 +262,7 @@ test-suite tests
     Hydra.Chain.Direct.Contract.CollectCom
     Hydra.Chain.Direct.Contract.Commit
     Hydra.Chain.Direct.Contract.Contest
+    Hydra.Chain.Direct.Contract.ContractGenerators
```
Often such modules are just called `.Gen`

v0d1ch committed:
Remove redundant import

v0d1ch committed:
Reduce the number of fanout outputs to 39 :(

v0d1ch committed:
Rename ContractGenerators to Gen
